### ISSUE TITLE
copy and paste using Ctrl-C Ctrl-V or right click menu

### ISSUE THREAD

**Issue by zhangfuwen** [termux/termux-app#1891]
**Feature description**

I am using this app(it is awesome) in desktop mode. I have a keyboard and mouse. I noticed that I can't paste to or copy texts from this app which is not convenient.

**Reference implementation**

I was not able to find a reference implementation.

**Comment by xeffyr** [comment]
> I can't paste

`CTRL + ALT + V` for paste.

Copying is not supported at the moment as Termux was designed for touch screen devices primarily.

**Comment by freebrowser1** [comment]
Weird.
When I log into Termux with SSH from a Mac using the standard Terminal app, copying just works like copying on the local terminal (Cmd+C) after selecting text. The same applies to a Linux terminal (Ctrl+Shift+C).

Windows is a different story, as copying and pasting into a Windows 'terminal' ('command prompt') uses some weird gestures and shortcuts.

From the Android device itself, one can hold a finger inside the Termunx terminal and Termux allows copying as well.

**Comment by coltexx** [comment]
👀

**Comment by zhangfuwen** [comment]
> Weird.
> When I log into Termux with SSH from a Mac using the standard macOS Terminal app, copying just works like copying on the local terminal (Cmd+C) after selecting text. The same applies to a Linux terminal (Ctrl+Shift+C), regardless whether Centos or Ubuntu / Debian.
> 
> Windows is a different story, as copying and pasting into a Windows 'terminal' ('command prompt') uses some weird gestures and shortcuts.
> 
> From the Android device itself, one can hold a finger inside the Termunx terminal and Termux allows copying as well.

I am using android device itself as a computer. So I am trying to copy and paste using termux app itself.

**Comment by ds6** [comment]
@freebrowser1
> Weird.

This is a discussion about the Termux terminal emulator—a part of and not the whole of Termux. Meanwhile you are connecting to Termux SSH daemon using other terminal emulators (Mac, Linux, Windows), which have different features. It's not weird at all that completely seperate terminal emulators would have completely independent feature sets...

@zhangfuwen Ctrl+[CV] are Windows conventions, in traditional Linux terminals Shift-Insert is used for pasting and selecting automatically copies to the buffer... So we should also be asking if Shift-Insert could be supported in this manner. I reflexively try to use it on Android still with the OSK to no effect. (This could also be due to strange keycodes being passed from Hacker's Keyboard if it otherwise should be working.)

**Comment by suhan-paradkar** [comment]
@agnostic-apollo @xeffyr I guess it will be more practical and easier to implement a `PASTE` extra-key to paste the clipboard content....

**Comment by agnostic-apollo** [comment]
@suhan-paradkar That can technically be implemented but what for use case do you want to use it for?

With hardware keyboard, `ctrl` + `alt` + `v` should work.

With touch screen, context menu is used for pasting, but requires long hold first, do you want this to be easier?

**Comment by suhan-paradkar** [comment]
> With hardware keyboard, `ctrl` + `alt` + `v` should work.

Yes.. But Extra keys hardly is required by those using hard keyboard

> With touch screen, context menu is used for pasting, but requires long hold first, do you want this to be easier?

Yes.. obviously.. It becomes hectic to first long press and then press upon the context menu when using multiple copy paste operations....

Also.. it is difficult to implement keyboard's own clipboard popup view...

It's easier.. both for implementation and usage.. a paste button to paste the clipboard content directly...

**Comment by agnostic-apollo** [comment]
Okay, will look into adding the extra key.

>Also.. it is difficult to implement keyboard's own clipboard popup view...

Do you mean termux's own clipboard with history like Gboard has?

**Comment by suhan-paradkar** [comment]
> Do you mean termux's own clipboard with history like Gboard has?

No.. I meant this overlay.... 
![Screenshot_2021-07-18-17-36-35-23](https://user-images.githubusercontent.com/76830230/126066382-5e51bb76-a699-40ef-be26-54f1a1d86e9d.jpg)

**Comment by agnostic-apollo** [comment]
That's only shown with https://wiki.termux.com/wiki/Touch_Keyboard#Text_Input_View

Keyboard suggestions, autocorrect, autocomplete etc do not work in normal mode since terminal view is not a text field.

**Comment by suhan-paradkar** [comment]
Yes ... That's why adding a PASTE button as aan extra key helps a lot...

**Comment by agnostic-apollo** [comment]
Okay.

**Comment by agnostic-apollo** [comment]
@suhan-paradkar Check 7d76e8b1

Moreover, `Ctrl`+`C` can't be used for copying since its supposed to send `SIGINT` and `Ctrl` + `Alt` + `C` can't be used since "someone" decided to use it for creating new termux sessions. https://wiki.termux.com/wiki/Hardware_Keyboard

**Comment by suhan-paradkar** [comment]
Great 👍👍 I will now add that to T: Wayland WIP Branch also....

**Comment by RalfWerner** [comment]
> Check 7d76e8b

@agnostic-apollo Unfortunately, I had _no success_ in the shot with your update of `termux.apk` and `nano .termux/*.Properties`:
![grafik](https://user-images.githubusercontent.com/45426380/126277500-c5846187-2c24-433a-b192-a373342482f9.png)
I've expected with `Copy` in the pop menu to get the selected text in the clipboard and the symbol and not the _word_ `PASTE` and clipboard text_ and not the _word._ Behind it is `Paste` from the _pop menu_ - correct.
PASTE in Extra Keys ist the first _and_ in popup over TAB. Both dont do it - **what is wrong and what you've got?**_

**Comment by agnostic-apollo** [comment]
The `PASTE` extra keys only pastes the current text in clipboard at current cursor position just like the `Paste` option from `Copy/Paste/More` menu. Working on android 7 device and android 10 avd.

**Comment by RalfWerner** [comment]
> The PASTE extra key only pastes the current text in clipboard at current cursor position

That is what I've exspected but not get (first 5 char above _red line_)! In `*properties` (Shot above) you find, what I've used. The Example is Android 11 and just now I've repeat it thth Androit 8 with the same result. _What **get you**. when **you doing the same?**_

**Comment by agnostic-apollo** [comment]
@suhan-paradkar is it working for you?

**Comment by suhan-paradkar** [comment]
I haven't checked it out yet...

**Comment by agnostic-apollo** [comment]
Cool, let me know if there is a problem.

**Comment by suhan-paradkar** [comment]
@agnostic-apollo I checked it in my device and it works perfectly.... 

> The idea is good but would it be not easier for T: X11 to use the same sources?_
> That would probably also simplify my commit to the Extra Key merge of the two apps.

@RalfWerner 
Well.. I actually use the same source.. but T:X11 implements its keyboard functioning in a slightly different way.. the reporting of keyboard events is mediated by libxkbcommon native library.... So.. same source... Slight changes

**Comment by agnostic-apollo** [comment]
Thanks. In future I may move extra keys to `termux-shared` if possible, but some `termux-app` specific logic exists inside it currently so will have to see if its possible easily. Probably require some interface too.

**Comment by suhan-paradkar** [comment]
@agnostic-apollo I guess it would be better if Extra-keys were moved into a seperate library instead of termux-shared... That way.. those devs who need extra-keys in their apps will need to just import a so-called "termux-extra-keys" library.. instead of whole of termux-shared... Which.. only some of them would want to be in their apps

> but some termux-app specific logic exists inside it currently so will have to see if its possible easily. Probably require some interface too.

Yes.. I have seen them... And had replaced them with T:X11 ones in one my fork's branches...

**Comment by agnostic-apollo** [comment]
>moved into a seperate library instead of termux-shared..

That would be more appropriate, lot of `termux-shared` stuff can be used by other apps and may be more appropriate in separate libraries. The issues is that then each library needs to be published separately, with `Jitpack` it should be relatively easy, although I'm not as complacent about it from a security point of view. Also versioning has to be done for each so will have to wait for a full release each time of libraries even if new stuff has been merged, unless instead of setting a library version in dependency, the commit hash is used. https://jitpack.io/docs/

>And had replaced them with T:X11 ones in one my fork's branches...

Yeah, saw that a while back. But if the library provides a base implementation, the importer can inherit the class and override relevant functions only to handle specific logic and then call the super method if needed to handle the common ones. It's doable but other changes need to be made to extra keys like locking the special extra keys on quick double tap or `Shift` support, so further design changes may be needed. I likely won't have time to do it in the near future though, other priorities.

**Comment by suhan-paradkar** [comment]
Great👍

**Comment by RalfWerner** [comment]
> @suhan-paradkar I checked it in my device and it works perfectly...

I suspect you cloned the current Termux app. Are your extra-keys in `*Properties` [comparable with my shot](https://github.com/termux/termux-app/issues/1891#issuecomment-883163490) or dit you check _T:X11_?
> Well.. I actually use the same source.. but T:X11 implements its keyboard ...

I check a [T:X11](https://github.com/suhan-paradkar/termux-wayland) clone from you (9 days old) where are the newest one (works perfectly)?

**Comment by suhan-paradkar** [comment]
> I check a T:X11 clone from you (9 days old) where are the newest one (works perfectly)?

I said that the Termux-app works perfectly.. not my fork....
I am yet to add Extra-keys to my fork...

**Comment by RalfWerner** [comment]
> I said that the Termux-app works perfectly.. not my fork...

Nice for you and `*Properties` comparable (! assume we use the same Termux-app)?

**Comment by RalfWerner** [comment]
If you use the last artifact of @twaik, the Extra-keys managed in an other storage than Termux see [here](https://github.com/termux/termux-x11/issues/54#issuecomment-1462338168) but works in the same way as before.

**Comment by lattice0** [comment]
control alt v does not work for me on Android Samsung Dex with hardware keyboard

**Comment by bkdwt** [comment]
> control alt v does not work for me on Android Samsung Dex with hardware keyboard

Same problem here when I try to past on my EC2 instance using Termux on Dex.

**Comment by lattice0** [comment]
@bkdwt termux should focus on Android for tablets as it's the next gen operating system. Something as simple as control v should work

**Comment by pzi123** [comment]
I manage to operate with the bluetooth mouse and keyboard connected to my samsung tablet running termux.  Left mouse button press/hold - just like the finger press - brings the cut/paste function where I can move the start/stop arrows and copy the text. To paste you can use ctrl-alt-v or again press/hold the left mouse button and the paste function is activated.
